### PR TITLE
A0-3083: Refactor ticker for chain extensions

### PR DESCRIPTION
**Pull request by timorl**
# Description

These are changes that are needed to use ticker with the periodic requests for chain extensions which we will be implementing soon. In particular this _should not_ change any current behaviour of the node, thus it's just a refactor.

Comments on whether this interface is clear or not are very welcome.

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- I have added tests
- I have made corresponding changes to the existing documentation
- I have created new documentation

### CODE REVIEW

File: finality-aleph/src/sync/ticker.rs [hunk 1]
```diff
@@ -1,39 +1,37 @@
 use tokio::time::{sleep, Duration, Instant};
 
 /// This struct is used for rate limiting as an on-demand ticker. It can be used for ticking
-/// at least once `max_timeout` but not more than once every `min_timeout`.
-/// Example usage would be to use `wait` method in main select loop and `try_tick` whenever
-/// you would like to tick sooner in another branch of select.
+/// at most after `max_timeout` but no sooner than after `min_timeout`.
+/// It will always tick after `max_timeout` and never after that unless reset earlier.
```
Comment (kostekIV) [inline finality-aleph/src/sync/ticker.rs:5]:
```suggestion
/// It will always tick before `max_timeout` and never after that unless reset earlier.
```
?

Comment (timorl) [reply finality-aleph/src/sync/ticker.rs]: Ugh, that's also not what I meant, but trying to write what I meant turned out to bee too hard, so I just deleted this line.

The documentation next to the methods should be good enough.

**Review by kostekIV — COMMENTED** [review]
Could you restore `wait_and_reset` fn for usecases where the user doesn't want to remember about resetting the timeout? This is a much less error prone approach. Why is the extraction of the reset is needed?

File: finality-aleph/src/sync/ticker.rs (hunk 1 above)
Comment (kostekIV) [reply finality-aleph/src/sync/ticker.rs]: Yep, it was confusing as hell :P

File: finality-aleph/src/sync/ticker.rs [hunk 2]
```diff
@@ -1,39 +1,36 @@
...
-/// at least once `max_timeout` but not more than once every `min_timeout`.
-/// Example usage would be to use `wait` method in main select loop and `try_tick` whenever
-/// you would like to tick sooner in another branch of select.
+/// at most after `max_timeout` but no sooner than after `min_timeout`.
+/// Example usage would be to use the `wait` method in main select loop and
+/// `try_tick` whenever you would like to tick sooner in another branch of select,
+/// resetting whenever the rate limited action actually occurs.
 pub struct Ticker {
-    last_tick: Instant,
+    last_reset: Instant,
```
Comment (kostekIV) [inline finality-aleph/src/sync/ticker.rs:9]: while we are changing this could we also rename current_timeout to some better name :<

**Comment by timorl** [comment]
The extraction of reset is needed because we will now have "failed ticks", i.e. the ticker ticks, but it turns out we don't actually want to send any message. We don't want to reset the ticker in this situation, since if anything suggests that we now want to send a message, we want to tick again immediately.

Making `wait` into `wait_and_reset` (or just keeping the name `wait_and_tick`) was something I even considered, the only issue I have with it now is that we either add weird boilerplate or always reset twice when we reach `max_timeout`. If you think the latter is acceptable then I can definitely change it, it only seemed somewhat annoying to me. To be clear, then we wouldn't even need a separate `wait` method – all the uses we have for this have a "forced" action when `wait` passes, so we would reset always anyway.

**Comment by timorl** [comment]
Hm, changed significantly, especially since while addressing comments I noticed this wouldn't've been enough for the application it is supposed for. <_<"

Should be clearer what is happening now, although we have a double reset if `max_timeout` actually occurs.

File: finality-aleph/src/sync/ticker.rs [hunk 3]
```diff
@@ -1,13 +1,18 @@
 use tokio::time::{sleep, Duration, Instant};
 
+enum State {
```
Comment (kostekIV) [inline finality-aleph/src/sync/ticker.rs:3]: `Mode` seems better here

**Review by kostekIV — APPROVED** [review]

File: finality-aleph/src/sync/ticker.rs (hunk 3 above)
Comment (timorl) [reply finality-aleph/src/sync/ticker.rs]: Done.

**Review by woocash2 — APPROVED** [review]
lgtm

**Pull request merged**